### PR TITLE
ECOM-5031 filter out excluded_course_runs from the program API

### PR DESCRIPTION
**Pull request by schenedx**
@clintonb @rlucioni Please review
This is the PR to filter out excluded course runs from the program API

### CODE REVIEW

File: course_discovery/apps/api/serializers.py [hunk 1]
```diff
@@ -261,11 +261,37 @@ class ContainedCoursesSerializer(serializers.Serializer):
...
+
+
 class ProgramSerializer(serializers.ModelSerializer):
-    courses = CourseSerializer(many=True)
+    courses = serializers.SerializerMethodField()
     authoring_organizations = OrganizationSerializer(many=True)
     type = serializers.SlugRelatedField(slug_field='name', queryset=ProgramType.objects.all())
 
+    def get_courses(self, program):
+        courses = Program.objects.get(id=program.id).courses
```
Comment (clintonb) [inline course_discovery/apps/api/serializers.py]: `courses = program.courses`

Comment (schenedx) [inline course_discovery/apps/api/serializers.py]: Good point. Updated

File: course_discovery/apps/api/tests/test_serializers.py [hunk 2]
```diff
@@ -169,6 +169,86 @@ def test_data_url_none(self):
         self.assertEqual(serializer.data['marketing_url'], None)
 
 
+class ProgramCourseSerializerTests(TestCase):
+    def test_no_run(self):
+        """
+        Make sure if the courses has no run,
+        the serializer still works as expected
+        """
+        request = make_request()
```
Comment (rlucioni) [inline course_discovery/apps/api/tests/test_serializers.py]: You can move this and some other shared code below to a `setUp()` method.

File: course_discovery/apps/api/tests/test_serializers.py [hunk 3]
```diff
@@ -169,6 +169,86 @@ def test_data_url_none(self):
         self.assertEqual(serializer.data['marketing_url'], None)
 
 
+class ProgramCourseSerializerTests(TestCase):
+    def test_no_run(self):
+        """
+        Make sure if the courses has no run,
```
Comment (rlucioni) [inline course_discovery/apps/api/tests/test_serializers.py]: > Make sure that if a course has no runs...

File: course_discovery/apps/api/tests/test_serializers.py [hunk 4]
```diff
@@ -169,6 +169,86 @@ def test_data_url_none(self):
...
+        program = ProgramFactory(courses=course_list)
+        serializer = ProgramCourseSerializer(
+            course_list,
+            many=True,
+            context={'request': request, 'program': program}
+        )
+
+        expected = CourseSerializer(course_list, many=True, context={'request': request}).data
+
+        self.assertSequenceEqual(serializer.data, expected)
```
Comment (rlucioni) [inline course_discovery/apps/api/tests/test_serializers.py:191]: Can you just use `assertEqual()`?

File: course_discovery/apps/api/tests/test_serializers.py [hunk 5]
```diff
@@ -169,6 +169,86 @@ def test_data_url_none(self):
...
+
+    def test_with_exclusions(self):
+        """
+        Test serializer with course_run exclusions within program
+        """
+        request = make_request()
+        course = CourseFactory()
+        image = course.image
+        video = course.video
+        exclused_runs = []
```
Comment (rlucioni) [inline course_discovery/apps/api/tests/test_serializers.py]: `excluded_runs`

File: course_discovery/apps/api/tests/test_serializers.py [hunk 6]
```diff
@@ -169,6 +169,86 @@ def test_data_url_none(self):
...
+            'marketing_url': '{url}?{params}'.format(
+                url=course.marketing_url,
+                params=urlencode({
+                    'utm_source': request.user.username,
+                    'utm_medium': request.user.referral_tracking_id,
+                })
+            )
+        }
+
+        self.assertDictEqual(serializer.data, expected)
```
Comment (rlucioni) [inline course_discovery/apps/api/tests/test_serializers.py:245]: I'd just use `assertEqual()`.

File: course_discovery/apps/api/tests/test_serializers.py [hunk 7]
```diff
@@ -187,7 +267,49 @@ def test_data(self):
...
+        """
+        Verify we can specify program excluded_course_runs and the serializers will
+        render the course_runs with exclusions
+        """
+        request = make_request()
+        org_list = OrganizationFactory.create_batch(1)
+        course_list = CourseFactory.create_batch(4)
+        for course in course_list:
+            course_runs = CourseRunFactory.create_batch(3, course=course)
+            excluded_runs = [course_runs[0]]
```
Comment (rlucioni) [inline course_discovery/apps/api/tests/test_serializers.py]: Assigning to this variable repeatedly within a for loop is strange. What are you trying to do here?

File: course_discovery/apps/api/tests/test_serializers.py [hunk 8]
```diff
@@ -169,6 +169,86 @@ def test_data_url_none(self):
...
+        self.assertSequenceEqual(serializer.data, expected)
+
+    def test_with_exclusions(self):
+        """
+        Test serializer with course_run exclusions within program
+        """
+        request = make_request()
+        course = CourseFactory()
+        image = course.image
+        video = course.video
```
Comment (rlucioni) [inline course_discovery/apps/api/tests/test_serializers.py]: Nitpick: you don't really need these `image` and `video` variables; you can lose a few lines by just accessing the properties.

**Comment by rlucioni** [comment]
@schenedx looks nice, just left a handful of notes inline.

File: course_discovery/apps/api/tests/test_serializers.py (hunk 4 above)
Comment (schenedx) [inline course_discovery/apps/api/tests/test_serializers.py:191]: I feel like using `assertSequenceEqual` is more accurate here

File: course_discovery/apps/api/tests/test_serializers.py (hunk 6 above)
Comment (schenedx) [inline course_discovery/apps/api/tests/test_serializers.py:245]: For other tests, we use `assertDictEqual`. It's more accurate and more consistent

File: course_discovery/apps/api/tests/test_serializers.py (hunk 3 above)
Comment (clintonb) [inline course_discovery/apps/api/tests/test_serializers.py]: Also, this can be on a single line.

**Comment by schenedx** [comment]
@rlucioni @clintonb I took in your review feedback. Please take another pass.

**Comment by rlucioni** [comment]
👍

**Pull request merged**